crates/ruff_linter/src/suppression.rs

```diff
@@ -145,13 +154,59 @@ impl Suppressions {
             return false;
         };
 
-        for suppression in &self.valid {
+        for suppression in &mut self.valid {
             if *code == suppression.code.as_str() && suppression.range.contains_range(range) {
+                suppression.used = true;
                 return true;
             }
         }
         false
     }
+
+    pub(crate) fn generate_diagnostics(&self, context: &LintContext, locator: &Locator) {
+        if !context.is_rule_enabled(Rule::UnusedSuppression) {
+            return;
+        }
+
+        let unused = self.valid.iter().filter(|suppression| !suppression.used);
+
+        for suppression in unused {
+            let Ok(rule) = Rule::from_code(&suppression.code) else {
+                continue; // TODO: invalid code
+            };
+            if !context.is_rule_enabled(rule) {
+                continue; // don't count disabled rules as unused
+            }
```

Comment:
I think this is a divergence from unused `noqa` suppressions, and I think it's very desirable to flag unused suppressions after you disabled a rule. That's why I think we shouldn't skip here